### PR TITLE
UserData (for VM and VMSS) Tests

### PR DESCRIPTION
**Pull request by roipmsft**
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/master/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/IVirtualMachineScaleSetsOperations.cs [hunk 1]
```diff
@@ -153,7 +157,7 @@ public partial interface IVirtualMachineScaleSetsOperations
         /// <exception cref="Microsoft.Rest.ValidationException">
         /// Thrown when a required parameter is null
         /// </exception>
-        Task<AzureOperationResponse<VirtualMachineScaleSet>> GetWithHttpMessagesAsync(string resourceGroupName, string vmScaleSetName, Dictionary<string, List<string>> customHeaders = null, CancellationToken cancellationToken = default(CancellationToken));
+        Task<AzureOperationResponse<VirtualMachineScaleSet>> GetWithHttpMessagesAsync(string resourceGroupName, string vmScaleSetName, string expand = default(string), Dictionary<string, List<string>> customHeaders = null, CancellationToken cancellationToken = default(CancellationToken));
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/IVirtualMachineScaleSetsOperations.cs:160]: This change requires a [Customization ](https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/compute/Microsoft.Azure.Management.Compute/src/Customizations) file. 

you will need to create a Customization file for this issue.
1 file in the Customization folder for each file in this PR that has a new parameter (not the last parameter) in the method signature.
Each Customization file needs these:

1) 1 method with each parameter up to but not including the new parameter.
2) 1 method for each parameter that is after the new parameter.
This means 1 new method for each parameter up to the parameter 'customHeaders' NOT including the new parameter.
Then 1 new method for each parameter up to 'cancellationToken' NOT including the new parameter.
Do this until you reach the last parameter.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/VirtualMachine.cs [hunk 2]
```diff
@@ -151,14 +151,17 @@ public VirtualMachine()
...
...ies), OSProfile osProfile = default(OSProfile), NetworkProfile networkProfile = default(NetworkProfile), SecurityProfile securityProfile = default(SecurityProfile), DiagnosticsProfile diagnosticsProfile = default(DiagnosticsProfile), SubResource availabilitySet = default(SubResource), SubResource virtualMachineScaleSet = default(SubResource), SubResource proximityPlacementGroup = default(SubResource), string priority = default(string), string evictionPolicy = default(string), BillingProfile billingProfile = default(BillingProfile), SubResource host = default(SubResource), SubResource hostGroup = default(SubResource), string provisioningState = default(string), VirtualMachineInstanceView instanceView = default(VirtualMachineInstanceView), string licenseType = default(string), string vmId = default(string), string extensionsTimeBudget = default(string), int? platformFaultDomain = default(int?), string userData = default(string), IList<VirtualMachineExtension> resources = default(IList<VirtualMachineExtension>), VirtualMachineIdentity identity = default(VirtualMachineIdentity), IList<string> zones = default(IList<string>), ExtendedLocation extendedLocation = default(ExtendedLocation))
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/VirtualMachine.cs:164]: Same Customization comment as above.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/VirtualMachineScaleSetVM.cs [hunk 3]
```diff
@@ -109,7 +112,7 @@ public VirtualMachineScaleSetVM()
...
...efault(bool?), string vmId = default(string), VirtualMachineScaleSetVMInstanceView instanceView = default(VirtualMachineScaleSetVMInstanceView), HardwareProfile hardwareProfile = default(HardwareProfile), StorageProfile storageProfile = default(StorageProfile), AdditionalCapabilities additionalCapabilities = default(AdditionalCapabilities), OSProfile osProfile = default(OSProfile), SecurityProfile securityProfile = default(SecurityProfile), NetworkProfile networkProfile = default(NetworkProfile), VirtualMachineScaleSetVMNetworkProfileConfiguration networkProfileConfiguration = default(VirtualMachineScaleSetVMNetworkProfileConfiguration), DiagnosticsProfile diagnosticsProfile = default(DiagnosticsProfile), SubResource availabilitySet = default(SubResource), string provisioningState = default(string), string licenseType = default(string), string modelDefinitionApplied = default(string), VirtualMachineScaleSetVMProtectionPolicy protectionPolicy = default(VirtualMachineScaleSetVMProtectionPolicy), string userData = default(string), Plan plan = default(Plan), IList<VirtualMachineExtension> resources = default(IList<VirtualMachineExtension>), IList<string> zones = default(IList<string>))
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/VirtualMachineScaleSetVM.cs:115]: Same customization comment as above.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/VirtualMachineUpdate.cs [hunk 4]
```diff
@@ -147,10 +147,13 @@ public VirtualMachineUpdate()
...
...dwareProfile), StorageProfile storageProfile = default(StorageProfile), AdditionalCapabilities additionalCapabilities = default(AdditionalCapabilities), OSProfile osProfile = default(OSProfile), NetworkProfile networkProfile = default(NetworkProfile), SecurityProfile securityProfile = default(SecurityProfile), DiagnosticsProfile diagnosticsProfile = default(DiagnosticsProfile), SubResource availabilitySet = default(SubResource), SubResource virtualMachineScaleSet = default(SubResource), SubResource proximityPlacementGroup = default(SubResource), string priority = default(string), string evictionPolicy = default(string), BillingProfile billingProfile = default(BillingProfile), SubResource host = default(SubResource), SubResource hostGroup = default(SubResource), string provisioningState = default(string), VirtualMachineInstanceView instanceView = default(VirtualMachineInstanceView), string licenseType = default(string), string vmId = default(string), string extensionsTimeBudget = default(string), int? platformFaultDomain = default(int?), string userData = default(string), VirtualMachineIdentity identity = default(VirtualMachineIdentity), IList<string> zones = default(IList<string>))
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/Models/VirtualMachineUpdate.cs:156]: Same customization comment as above.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/SdkInfo_ComputeManagementClient.cs [hunk 5]
```diff
@@ -43,14 +43,16 @@ internal static partial class SdkInfo
                 new Tuple<string, string, string>("Compute", "VirtualMachineExtensions", "2021-03-01"),
                 new Tuple<string, string, string>("Compute", "VirtualMachineImages", "2021-03-01"),
                 new Tuple<string, string, string>("Compute", "VirtualMachineImagesEdgeZone", "2021-03-01"),
-                new Tuple<string, string, string>("Compute", "VirtualMachineRunCommands", "2021-03-01"),
+                new Tuple<string, string, string>("Compute", "VirtualMachineRunCommands", "2020-12-01"),
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/SdkInfo_ComputeManagementClient.cs:46]: This is incorrect. You might be based off an older swagger branch. Undo this.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineRunCommandsOperations.cs [hunk 6]
```diff
@@ -94,7 +94,7 @@ public async Task<AzureOperationResponse<IPage<RunCommandDocumentBase>>> ListWit
             {
                 throw new ValidationException(ValidationRules.CannotBeNull, "this.Client.SubscriptionId");
             }
-            string apiVersion = "2021-03-01";
+            string apiVersion = "2020-12-01";
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineRunCommandsOperations.cs:97]: Same issue. Probably using an older version of the compute-2021-03 branch.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMRunCommandsOperations.cs [hunk 7]
```diff
@@ -201,7 +201,7 @@ public async Task<AzureOperationResponse<VirtualMachineRunCommand>> GetWithHttpM
             {
                 throw new ValidationException(ValidationRules.CannotBeNull, "this.Client.SubscriptionId");
             }
-            string apiVersion = "2021-03-01";
+            string apiVersion = "2020-12-01";
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMRunCommandsOperations.cs:204]: same.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMsOperations.cs [hunk 8]
```diff
@@ -3367,7 +3367,7 @@ public async Task<AzureOperationResponse<RunCommandResult>> BeginRunCommandWithH
             {
                 throw new ValidationException(ValidationRules.CannotBeNull, "this.Client.SubscriptionId");
             }
-            string apiVersion = "2021-03-01";
+            string apiVersion = "2020-12-01";
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMsOperations.cs:3370]: Using an older version of Swagger to generate this. Incorrect.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetsOperations.cs [hunk 9]
```diff
@@ -348,7 +352,7 @@ public async Task<AzureOperationResponse> DeleteWithHttpMessagesAsync(string res
         /// <return>
         /// A response object containing the response body and response headers.
         /// </return>
-        public async Task<AzureOperationResponse<VirtualMachineScaleSet>> GetWithHttpMessagesAsync(string resourceGroupName, string vmScaleSetName, Dictionary<string, List<string>> customHeaders = null, CancellationToken cancellationToken = default(CancellationToken))
+        public async Task<AzureOperationResponse<VirtualMachineScaleSet>> GetWithHttpMessagesAsync(string resourceGroupName, string vmScaleSetName, string expand = default(string), Dictionary<string, List<string>> customHeaders = null, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetsOperations.cs:355]: same customization comment as above.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetsOperationsExtensions.cs [hunk 10]
```diff
@@ -223,12 +227,16 @@ public static VirtualMachineScaleSet Get(this IVirtualMachineScaleSetsOperations
...
             /// </param>
+            /// <param name='expand'>
+            /// The expand expression to apply on the operation. Possible values include:
+            /// 'userData'
+            /// </param>
             /// <param name='cancellationToken'>
             /// The cancellation token.
             /// </param>
-            public static async Task<VirtualMachineScaleSet> GetAsync(this IVirtualMachineScaleSetsOperations operations, string resourceGroupName, string vmScaleSetName, CancellationToken cancellationToken = default(CancellationToken))
+            public static async Task<VirtualMachineScaleSet> GetAsync(this IVirtualMachineScaleSetsOperations operations, string resourceGroupName, string vmScaleSetName, string expand = default(string), CancellationToken cancellationToken = default(CancellationToken))
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetsOperationsExtensions.cs:237]: Same customization comment.

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperations.cs [hunk 11]
```diff
@@ -5038,7 +5038,7 @@ public async Task<AzureOperationResponse<RunCommandResult>> BeginRunCommandWithH
             {
                 throw new ValidationException(ValidationRules.CannotBeNull, "this.Client.SubscriptionId");
             }
-            string apiVersion = "2021-03-01";
+            string apiVersion = "2020-12-01";
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachinesOperations.cs:5041]: same issue as above.

File: sdk/compute/Microsoft.Azure.Management.Compute/tests/ScenarioTests/UserDataTests.cs [hunk 12]
```diff
@@ -0,0 +1,61 @@
+﻿using Microsoft.Azure.Management.Compute;
+using Microsoft.Azure.Management.Compute.Models;
+using Microsoft.Azure.Management.ResourceManager;
+using Microsoft.Rest.ClientRuntime.Azure.TestFramework;
+using System;
+using Xunit;
+
+namespace Compute.Tests
```
Comment (Sandido) [inline sdk/compute/Microsoft.Azure.Management.Compute/tests/ScenarioTests/UserDataTests.cs:8]: Please include the generated json file made by running this test in Record mode.

**Review by Sandido — CHANGES_REQUESTED** [review]

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/IVirtualMachineScaleSetsOperations.cs (hunk 1 above)
Comment (roipmsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/IVirtualMachineScaleSetsOperations.cs:162]: thank you for the explanation, but sorry, I'm having trouble understanding these directions.
Is there a wiki or some documentation that also explaints this?

you said: This means 1 new method for each parameter up to...
so if a method has ten parameters, does that means I would need ~ten new methods?

Comment (roipmsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/IVirtualMachineScaleSetsOperations.cs:162]: can we use a different example:
Let's say we have this method
Method(a, b, c, d, e, f)
now i am adding a new param h between d and e
Method(a, b, c, d, h, e, f)

what would the customization file require?

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/SdkInfo_ComputeManagementClient.cs (hunk 5 above)
Comment (roipmsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/SdkInfo_ComputeManagementClient.cs]: fixed

File: sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMsOperations.cs (hunk 8 above)
Comment (roipmsft) [reply sdk/compute/Microsoft.Azure.Management.Compute/src/Generated/VirtualMachineScaleSetVMsOperations.cs]: fixed

**Pull request closed without merging**